### PR TITLE
Controller Changes + Profile Completion

### PR DESCRIPTION
**Pull request by NotSalH**
This merge request allows the admin to add new users

### CODE REVIEW

File: src/main/java/medicaldoctor/controllers/navbar/LogOutController.java [hunk 1]
```diff
@@ -3,13 +3,15 @@
...
  */
-package medicaldoctor.controllers;
+package medicaldoctor.controllers.navbar;
 
 import java.net.URL;
 import java.util.ResourceBundle;
 import javafx.event.ActionEvent;
 import javafx.fxml.FXML;
 import javafx.fxml.Initializable;
+import medicaldoctor.controllers.ControllerManager;
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/controllers/navbar/LogOutController.java]: you have imports but i dont see any changes.
ctrl+shift+i to manage imports

File: src/main/java/medicaldoctor/controllers/navbar/LogOutController.java [hunk 2]
```diff
@@ -3,13 +3,15 @@
...
+import medicaldoctor.controllers.ParentController;
 import medicaldoctor.core.AppSession;
 
 public class LogOutController implements Initializable, ParentController{
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/controllers/navbar/LogOutController.java]: alt+shift+f to autoformat
that is not a one-time thing, it is a constant thing to keep everything clean.

Comment (tlacasse) [reply src/main/java/medicaldoctor/controllers/navbar/LogOutController.java]: on every file too, not just ones i comment on

File: src/main/java/medicaldoctor/controllers/navbar/ProfileController.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            username_tf.setText(uig.getEmail(first_name_tf.getText(), last_name_tf.getText()));
+            conformation.setVisible(true);
+        }
+        else{
+            no_field.setVisible(true);
+        }
+    }
+    
+    @FXML
+    void confirm_user(ActionEvent event){
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/controllers/navbar/ProfileController.java]: user registration is its own page, not on the profile.
also use `UserService.createNewUser`

File: src/main/java/medicaldoctor/controllers/navbar/NavigationBarController.java [hunk 4]
```diff
@@ -72,103 +81,123 @@
...
+    void redBloodCell(ActionEvent event) throws Exception{
+        AppSession.cm.load(LookUp.RED_BLOOD_CELLS);
+        AppSession.cm.showScreen(LookUp.RED_BLOOD_CELLS);
     }
     
     @FXML
-    void whiteBloodCell(ActionEvent event){
-        
+    void whiteBloodCell(ActionEvent event) throws Exception{
+        AppSession.cm.load(LookUp.WHITE_BLOOD_CELLS);
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/controllers/navbar/NavigationBarController.java]: you could have another method that calls both of these
it is excessive to call two methods with the same input every time.

File: src/main/java/medicaldoctor/controllers/navbar/NavigationBarController.java [hunk 5]
```diff
@@ -72,103 +81,123 @@
...
+    void patientRegistration(ActionEvent event) throws Exception{
+        AppSession.cm.load(LookUp.PATIENT_REGISTRATION);
+        AppSession.cm.showScreen(LookUp.PATIENT_REGISTRATION);
     }
     
     @FXML
-    void patientLookup(ActionEvent event){
-        
+    void patientLookup(ActionEvent event) throws Exception{
+        AppSession.cm.load(LookUp.PATIENT_LOOKUP);
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/controllers/navbar/NavigationBarController.java]: rename `cm` to `CONTROLLER_MANAGER` or something more descriptive. don't use abbreviations.

File: src/main/java/medicaldoctor/controllers/navbar/ProfileController.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+import javafx.scene.control.ChoiceBox;
+import javafx.scene.control.Label;
+import javafx.scene.control.TextField;
+import medicaldoctor.controllers.ControllerManager;
+import medicaldoctor.controllers.ParentController;
+import medicaldoctor.core.AppSession;
+import static medicaldoctor.core.AppSession.ENCRYPTION;
+import medicaldoctor.entities.User;
+
+public class ProfileController implements Initializable, ParentController{
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/controllers/navbar/ProfileController.java]: with static controller manager, you dont need the ParentController interface

File: src/main/java/medicaldoctor/controllers/navbar/ProfileController.java (hunk 3 above)
Comment (tlacasse) [reply src/main/java/medicaldoctor/controllers/navbar/ProfileController.java]: this is already implemented

Comment (tlacasse) [reply src/main/java/medicaldoctor/controllers/navbar/ProfileController.java]: and for future reference if it wasn't implemented you need to put db stuff in the DatabaseContext

File: src/main/java/medicaldoctor/controllers/navbar/ProfileController.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        employee_type_cb.getItems().add("Admin");
+        employee_type_cb.getItems().add("Executive");
+        employee_type_cb.getItems().add("Nurse");
+        employee_type_cb.getItems().add("Staff");
+        employee_type_cb.getItems().add("LabWorker");
+        uig = new UserInformationGenerator();
+    }
+
+    @Override
+    public void setScreenParent(ControllerManager page){
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/controllers/navbar/ProfileController.java]: with the static controller, you dont need this inherited method.

File: src/main/java/medicaldoctor/controllers/navbar/ProfileController.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    @FXML
+    void change_password(ActionEvent event){
+        
+    }
+    
+    @Override
+    public void initialize(URL location, ResourceBundle resources){
+        sex_cb.getItems().add("M");
+        sex_cb.getItems().add("F");
+        employee_type_cb.getItems().add("Doctor");
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/controllers/navbar/ProfileController.java]: doesn't this look like it could use a loop????
think about maintainability whenever writing anything, if we add a new usertype you would have to 
go everywhere there is something like this and make a change.

File: src/main/java/medicaldoctor/controllers/navbar/patientRegistrationController.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
+/*
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/controllers/navbar/patientRegistrationController.java]: you are still not removing these. once again just because it is autogenerated doesn't mean you are forced to keep it... go through all your files and clean them up.

Comment (tlacasse) [reply src/main/java/medicaldoctor/controllers/navbar/patientRegistrationController.java]: with autoformatting too

File: src/main/java/medicaldoctor/entities/UserInformationGenerator.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
+package medicaldoctor.entities;
+
+import java.util.Calendar;
+import java.util.Date;
+import java.util.Random;
+
+public class UserInformationGenerator {
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/entities/UserInformationGenerator.java]: remember you are working on controllers not the backend services

File: src/main/resources/PatientRegistration.fxml [hunk 11]
```diff
@@ -9,7 +9,7 @@
 <?import javafx.scene.control.*?>
 <?import javafx.scene.layout.*?>
 
-<BorderPane maxHeight="-Infinity" maxWidth="-Infinity" minHeight="-Infinity" minWidth="-Infinity" prefHeight="745.0" prefWidth="1136.0" xmlns="http://javafx.com/javafx/8" xmlns:fx="http://javafx.com/fxml/1" fx:controller="medicaldoctor.PatientRegistrationController">
+<BorderPane maxHeight="-Infinity" maxWidth="-Infinity" minHeight="-Infinity" minWidth="-Infinity" prefHeight="745.0" prefWidth="1136.0" xmlns="http://javafx.com/javafx/8" xmlns:fx="http://javafx.com/fxml/1" fx:controller="medicaldoctor.controllers.navbar.patientRegistrationController">
```
Comment (tlacasse) [inline src/main/resources/PatientRegistration.fxml]: i don't understand this `navbar package`.... there is only one navbar controller

File: src/main/java/medicaldoctor/controllers/navbar/ProfileController.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+        user.setType(uig.getType((String)employee_type_cb.getValue()));
+        user.setDepartment("IDK");
+        user.setOfficeNum(Integer.parseInt(office_tf.getText()));
+        user.save();
+        //Transistion to another page
+    }
+    
+    @FXML
+    void change_password(ActionEvent event){
+
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/controllers/navbar/ProfileController.java]: change password service is implemented

File: src/main/java/medicaldoctor/controllers/navbar/ProfileController.java [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    Label no_field;
+    
+    @FXML
+    Button conformation;
+    
+    UserInformationGenerator uig;
+    
+    @FXML
+    void new_user(ActionEvent event){
+        if(!(last_name_tf.getText().isEmpty() && first_name_tf.getText().isEmpty() && phone_tf.getText().isEmpty() && office_tf.getText().isEmpty())){
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/controllers/navbar/ProfileController.java]: use loops.

**Review by tlacasse — CHANGES_REQUESTED** [review]

File: src/main/resources/PatientRegistration.fxml (hunk 11 above)
Comment (NotSalH) [reply src/main/resources/PatientRegistration.fxml]: anything that is in the navbar goes there. It's to clean things up for me

File: src/main/java/medicaldoctor/controllers/navbar/ProfileController.java (hunk 12 above)
Comment (NotSalH) [reply src/main/java/medicaldoctor/controllers/navbar/ProfileController.java]: Then why is it in the FXML file? it's a button it REQUIRES this method.

File: src/main/java/medicaldoctor/entities/UserInformationGenerator.java (hunk 10 above)
Comment (NotSalH) [reply src/main/java/medicaldoctor/entities/UserInformationGenerator.java]: I needed a way to get information such as email. I didn't have it so I made it. I did it so I can test the controller.

File: src/main/java/medicaldoctor/controllers/navbar/NavigationBarController.java (hunk 4 above)
Comment (NotSalH) [reply src/main/java/medicaldoctor/controllers/navbar/NavigationBarController.java]: I don't know what you mean. I'm using whatever is provided for me with the FXML there are two buttons. They will have two different actions. Otherwise have one button that says BloodWork or something.

File: src/main/java/medicaldoctor/controllers/navbar/ProfileController.java (hunk 6 above)
Comment (NotSalH) [reply src/main/java/medicaldoctor/controllers/navbar/ProfileController.java]: We don't have a lot of time. If I spend all my time making fixes like this we won't get anything done. This works so please leave it

File: src/main/java/medicaldoctor/controllers/navbar/ProfileController.java (hunk 7 above)
Comment (NotSalH) [reply src/main/java/medicaldoctor/controllers/navbar/ProfileController.java]: We don't have a lot of time. If I spend all my time making fixes like this we won't get anything done. This works so please leave it

File: src/main/java/medicaldoctor/controllers/navbar/NavigationBarController.java (hunk 4 above)
Comment (tlacasse) [reply src/main/java/medicaldoctor/controllers/navbar/NavigationBarController.java]: i mean like 
```java
AppSession.cm.loadAndShow(Lookup.XXX)
```

File: src/main/java/medicaldoctor/controllers/navbar/ProfileController.java (hunk 6 above)
Comment (tlacasse) [reply src/main/java/medicaldoctor/controllers/navbar/ProfileController.java]: I agree, but I am providing comments so they can be changed when first time implementing other controllers.

Comment (tlacasse) [reply src/main/java/medicaldoctor/controllers/navbar/ProfileController.java]: We can learn how to fix things for the future so we aren't continually producing a messy change.

File: src/main/java/medicaldoctor/controllers/navbar/ProfileController.java (hunk 7 above)
Comment (tlacasse) [reply src/main/java/medicaldoctor/controllers/navbar/ProfileController.java]: see above comment, but for example these, you simply don't include the interface

Comment (tlacasse) [reply src/main/java/medicaldoctor/controllers/navbar/ProfileController.java]: Leave the ones now, don't continue in the future thats what I am saying.

File: src/main/java/medicaldoctor/entities/UserInformationGenerator.java (hunk 10 above)
Comment (tlacasse) [reply src/main/java/medicaldoctor/entities/UserInformationGenerator.java]: If you recall in one of our discord convos, controllers should have code that doesn't need to be tested.

File: src/main/resources/PatientRegistration.fxml (hunk 11 above)
Comment (tlacasse) [reply src/main/resources/PatientRegistration.fxml]: anything that is in the navbar is not a class thus is not in a package.
patient registration is not in the navbar either????

File: src/main/java/medicaldoctor/controllers/navbar/ProfileController.java (hunk 12 above)
Comment (tlacasse) [reply src/main/java/medicaldoctor/controllers/navbar/ProfileController.java]: change password **service** is implemented. just letting you know so you can fill in this controller.

File: src/main/java/medicaldoctor/controllers/LogOutController.java [hunk 14]
```diff
@@ -26,6 +26,7 @@ void exitButtonClicked(ActionEvent event){
     @FXML
     void loginPageButtonClicked(ActionEvent event) throws Exception{
         AppSession.CONTROLLER_MANAGER.showScreen(LookUp.LOGIN_SCREEN);
+        //This does not logout user in backend.
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/controllers/LogOutController.java]:
```suggestion
        AppSession.setActiveUser(null);
```

Comment (tlacasse) [reply src/main/java/medicaldoctor/controllers/LogOutController.java]: preferrably would create a logout method on the appsession that will do that plus clear the other fields such as the current visit and current patient

File: src/main/java/medicaldoctor/controllers/LookUp.java [hunk 15]
```diff
@@ -34,4 +38,12 @@
     public static final String CT = "Ct";
     public static final String MRI = "Mri";
     public static final String PROFILE = "Profile";
+    
+    public static ObservableList<String> STATES = FXCollections.observableArrayList(
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/controllers/LookUp.java]: cool

File: src/main/java/medicaldoctor/controllers/PatientRegistrationController.java [hunk 16]
```diff
@@ -0,0 +1,68 @@
+/*
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/controllers/PatientRegistrationController.java]: still

File: src/main/java/medicaldoctor/controllers/LookUp.java (hunk 15 above)
Comment (tlacasse) [reply src/main/java/medicaldoctor/controllers/LookUp.java]: good

File: src/main/java/medicaldoctor/controllers/PatientRegistrationController.java [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+    
+    @FXML
+    TextField zipCodeMailingAddressNumber, textStreetBillingAddress, textCityBillingAddress, zipCodeBillingAddressNumber;
+    
+    @FXML
+    TextField textDoctor, textVisitDate, textAreaChiefComplaint, textAreaPresentIllness;*/
+    
+    @FXML
+    void submitButtonClick(ActionEvent event){
+
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/controllers/PatientRegistrationController.java]: put patient registration stuff into a different PR, try to limit adding and adding to PRs

File: src/main/resources/LogOut.fxml [hunk 18]
```diff
@@ -18,3 +22,9 @@
       <Button fx:id="loginPageButton" layoutX="329.0" layoutY="215.0" mnemonicParsing="false" onAction="#loginPageButtonClicked" prefHeight="25.0" prefWidth="112.0" text="Log-In Page" />
    </children>
 </AnchorPane>
+
+   </center>
+   <top>
+       <fx:include fx:id="navigationBar" source="NavigationBar.fxml" />
```
Comment (tlacasse) [inline src/main/resources/LogOut.fxml]: logout does not have navbar

**Review by tlacasse — CHANGES_REQUESTED** [review]

File: src/main/java/medicaldoctor/entities/User.java [hunk 19]
```diff
@@ -39,7 +39,10 @@
 
     @Column(name = "UserName", nullable = false, length = 50, unique = true)
     private String userName;
-
+    
+    @Column(name = "Email", nullable = false, length = 50, unique = true)
+    private String email;
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/entities/User.java]: please look at what  you are copying, this is the third time you are  adding a entity field and not removing the "unique".
I admit this *is* different, because emails should be unique, but we don't need an index on this field

File: src/main/java/medicaldoctor/entities/User.java [hunk 20]
```diff
@@ -85,7 +88,11 @@ public String getLastName() {
     public void setLastName(String lastName) {
         this.lastName = lastName;
     }
-
+    
+    public void setEmail(){
+        this.email = this.firstName.substring(0, 1) + this.lastName.substring(1, this.lastName.length());
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/entities/User.java]: put this in a EmailGenerator class like the password and username, we dont want to hardcode this here.

Comment (tlacasse) [reply src/main/java/medicaldoctor/entities/User.java]: also make it an actual email :)

File: src/main/java/medicaldoctor/entities/User.java (hunk 19 above)
Comment (tlacasse) [reply src/main/java/medicaldoctor/entities/User.java]: use `unique` only if we are commonly querying by this column, like usernames

**Review by tlacasse — CHANGES_REQUESTED** [review]

File: src/main/java/medicaldoctor/util/EmailNameGenerator.java [hunk 21]
```diff
@@ -0,0 +1,10 @@
+package medicaldoctor.util;
+
+public class EmailNameGenerator implements EmailGenerator{
+    
+    @Override
+    public String generateEmail(String first_name, String last_name) {
+        return first_name.substring(0, 1) + last_name.substring(1, last_name.length()) + "@MCARE.com";
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/util/EmailNameGenerator.java]: how about just passing in the username.
also add email generator to createNewUser service

File: src/main/java/medicaldoctor/setup/InitTestDatabase.java [hunk 22]
```diff
@@ -103,7 +103,6 @@ private static void insertTestRecords() {
         user.setType(UserType.ADMIN);
         user.setDepartment("Administration");
         user.setOfficeNum(145);
-        user.setEmail();
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/setup/InitTestDatabase.java]: it should still be set, you can have a local EmailGenerator to set the values,
maybe even put a loop at the end:
```java
for(User user : somelist){
    user.setEmail(emailGenerator.generateEmail(user.getUsername()));
}
```

File: src/main/java/medicaldoctor/core/AppSession.java [hunk 23]
```diff
@@ -20,7 +22,10 @@
     public static UserNameGenerator USER_NAME_GENERATOR
             = new FirstLetterLastNameUserNameGenerator();
     public static Encryption ENCRYPTION = new Encryption();
-
+    
+    public static EmailGenerator EMAIL_GENERATOR
+            = new EmailNameGenerator();
```
Comment (tlacasse) [inline src/main/java/medicaldoctor/core/AppSession.java]: use this in UserService::createNewUser and add to the unit tests.

**Review by tlacasse — CHANGES_REQUESTED** [review]
